From 163eab1e19877281b92d4d9555f8215a649920c1 Mon Sep 17 00:00:00 2001
From: Amelia Crate <amelia.crate@chainguard.dev>
Date: Tue, 29 Jul 2025 12:06:47 -0700
Subject: [PATCH 3/3] crypto: reusable jitterentropy

modify crypto/Makefile for highly reusable jitterentropy

* include vendored sha3_jitterentropy in jitterentropy_rng module
* include module and shim objects in jitterentropy_rng_module
* use prebuilt jitterentropy and sha3 objects unless REBUILD_JITTERENTROPY is set
* disable UBSAN, KASAN, KCOV_INSTRUMENT for objects which should be prebuilt

Guarding the CFLAGS stuff not necessary, but is nice because jitterentropy emits #error if
compiled with optimizations, so this is a small guardrail against unintentional
rebuilding of jitterentropy.
---
 crypto/Makefile | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/crypto/Makefile b/crypto/Makefile
index 017df3a2e4bb..050774163cdb 100644
--- a/crypto/Makefile
+++ b/crypto/Makefile
@@ -3,6 +3,15 @@
 # Cryptographic API
 #
 
+jent_sources = jitterentropy sha3_jitterentropy
+# Use 'make REBUILD_JITTERENTROPY=1' to ignore prebuilt objects
+ifdef REBUILD_JITTERENTROPY
+$(foreach jent_src,$(jent_sources),$(eval $(obj)/$(jent_src).o: $(src)/$(jent_src).c))
+else
+$(foreach jent_src,$(jent_sources),$(obj)/$(jent_src).o): $(subst $(obj),$(src),$(@))
+	$(call if-changed,copy)
+endif
+
 obj-$(CONFIG_CRYPTO) += crypto.o
 crypto-y := api.o cipher.o
 
@@ -169,10 +178,17 @@ obj-$(CONFIG_CRYPTO_RNG2) += rng.o
 obj-$(CONFIG_CRYPTO_ANSI_CPRNG) += ansi_cprng.o
 obj-$(CONFIG_CRYPTO_DRBG) += drbg.o
 obj-$(CONFIG_CRYPTO_JITTERENTROPY) += jitterentropy_rng.o
+ifdef REBUILD_JITTERENTROPY
 CFLAGS_jitterentropy.o = -O0
+# Disable for certified objects
 KASAN_SANITIZE_jitterentropy.o = n
 UBSAN_SANITIZE_jitterentropy.o = n
-jitterentropy_rng-y := jitterentropy.o jitterentropy-kcapi.o
+KCOV_INSTRUMENT_jitterentropy.o = n
+KASAN_SANITIZE_sha3_jitterentropy.o = n
+UBSAN_SANITIZE_sha3_jitterentropy.o = n
+KCOV_INSTRUMENT_sha3_jitterentropy.o = n
+endif
+jitterentropy_rng-y := jitterentropy.o jitterentropy-kcapi.o jitterentropy-shim.o sha3_jitterentropy.o
 obj-$(CONFIG_CRYPTO_JITTERENTROPY_TESTINTERFACE) += jitterentropy-testing.o
 obj-$(CONFIG_CRYPTO_BENCHMARK) += tcrypt.o
 obj-$(CONFIG_CRYPTO_GHASH) += ghash-generic.o
-- 
2.43.0

